[#35] OrderService unit test

jjeda opened the pull request:
- 정상적인 주문 생성 단위테스트
- 주문관련 도메인 Getter 에 @JsonIgnore 처리

jjeda committed:
[#35] OrderService unit test

- 정상적인 주문 생성 단위테스트
- 주문관련 도메인 Getter 에 @JsonIgnore 처리

jjeda commented on src/test/java/me/jjeda/mall/orders/service/OrderServiceTest.java:
```diff
@@ -0,0 +1,112 @@
+        assertThat(createOrder).isNotNull();
+        assertThat(createOrder.getAccount()).isEqualTo(buyer);
+        assertThat(createOrder.getDelivery().getOrder()).isEqualTo(createOrder);
+        assertThat(createOrder.getOrderItems().get(0).getOrder()).isEqualTo(createOrder);
```
104번 라인이 실행되면 마지막에 `orderRepository.save(order);` 가 실행되기 때문에 결국 101번 라인이 실행되는 것으로 이해하였습니다. 따라서 108번 라인은 문제없이 통과가 되는데 109번, 110번 라인은 통과되지 않습니다. (100번 라인처럼 `order`에 처리해주지 않았기 때문에) 

제가 이 테스트를 통해 확인하고 싶은 것은 `OrderService` 객체의 `createOrder` 메서드 내부 로직이 잘 동작하는지 확인하기 위한 테스트인데(dto -> entity) 100번라인처럼 (109,110번 라인도) 설정해주고 return 받는다면 결국 `orderRepository.save()` 가 호출 되었는가? 만 확인하는 테스트 인것 같습니다.. 

질문을 정리하면
`orderDto` 만으로 서비스의 로직을 수행하고 `createOrder` 를 봤을 때 내부 로직이 잘 동작 되었는지 확인하는 방법이 궁금합니다.

jjeda replied:
오늘 계속 찾아봤는데 `ArgumentMather` 를 implement 하면 내부 argument 값을 비교할 수 있는걸 찾았습니다

```java
 //then
        verify(orderRepository, times(1)).save(
                argThat(order1 -> order1.getAccount().equals(buyer)
                        && Objects.nonNull(order1.getDelivery().getOrder())
                        && Objects.nonNull(order1.getOrderItems().get(0).getOrder())));
```

변경 된 코드인데 `buyer` 는 파라미터로 넘기기때문에 비교가 가능했는데 배송에 `order` 가 들어갔는지 주문상품들에 주문정보가 들어갔는지 (Null이 아님) 만 확인이 가능했습니다.. 

이 정도로 테스트해도 되는지 아니면 다른 방법이 있을까요~?

jjeda replied:
``` java

//then
verify(orderRepository, times(1)).save(
    argThat(innerOrder -> innerOrder.getAccount().equals(buyer)
        && innerOrder.getDelivery().getOrder().equals(innerOrder)
        && innerOrder.getOrderItems().get(0).getOrder().equals(innerOrder)));

```

최종적으로 이렇게 바꿔봤습니다. 적절한지 궁금하네요..

f-lab-dev replied:
음 소스를 살펴봤는데요. 테스트케이스가 다 반영되지 않는것같네요.
테스트하기 좋은 소스가 좋은 소스다라는 말이 있습니다.

orderDto는 메소드 내부에서 보면 toEntity()만 호출하고 사용하지 않습니다.
그래서 이 값을 쓸 수가 없는거죠.

아래처럼 해주면 보다 수월해질 것 같습니다.

1. OrderDto를 정의하기보다는 OrderDto가 리턴할 엔티티 객체를 정의해주고,
2. OrderDto를 Mock으로 만든다음,
3. given을 사용해 toEntity() 메소드에서 1번의 객체를 리턴하게 만들어주고
4. `orderRepository.save()`에서 해당 엔티티를 리턴하게 해줍니다.

f-lab-dev replied:
그리고 부가적으로 Account 엔티티를 받는것도 개선하는게 좋을 것 같네요. 이것도 의존성이 너무 퍼져있어서 잘못 사용하게 될 가능성이 보입니다.

jjeda committed:
[#35] OrderService Unit Test

- 기존 단위테스트 수정
- 서비스 레이어 단위테스트 보완

f-lab-dev commented on src/test/java/me/jjeda/mall/orders/service/OrderServiceTest.java:
```diff
@@ -0,0 +1,205 @@
+    @TestDescription("없는 주문정보를 불러올때 예외처리하는 테스트")
+    public void getOrder_Throw_Entity_Not_Found() {
+        //given
+        Order order = orderDto.toEntity();
```
이 order는 사용하지 않는 것 같습니다.

f-lab-dev reviewed (commented):
나머지는 비즈니스로직이 그래도 괜찮게 검증되고 있는 것 같네요

jjeda committed:
[#35] AccountService 유닛테스트 수정

- 정상적으로 주문생성하는 테스트 수정
- AccountController -> AccountService 에서 Entity 가아닌 Dto를 받게 수정

jjeda committed:
test automation 테스트

f-lab-dev commented on src/test/java/me/jjeda/mall/orders/service/OrderServiceTest.java:
```diff
@@ -54,72 +57,78 @@
 
         //given
-        given(orderRepository.save(any())).willReturn(orderDto.toEntity());
+        given(orderDto.toEntity()).willReturn(order);
```
👍

jjeda committed:
[#13] Account 은닉화

- AccountService 에서 Entity가 아닌 Dto를 리턴하도록 수정
- AccountController에서 Entity 를 받아오는 것이 아닌 Dto를받아오게 수정
- Account <-> AccountDto Circular Dependency 문제해결을 위한 AccountAndDtoAdapter 적용
- Session에 저장할때 Account가 아닌 AccountDto를 저장하도록 수정
- 현재사용자를 받아올때 account를 받아오는것이아닌 AccountDto를 받아오도록 수정
- 테스트코드 수정

jjeda committed:
[#13] Account <-> Dto 어댑터 수정

- DtoToEntity 메서드 -> dtoToEntity
- 메서드 중복 로직 제거

jjeda committed:
Merge pull request #37 from f-lab-edu/feature/13

[#13] Account 은닉화
